module/zfs/spa.c

```diff
@@ -9589,6 +9602,164 @@ spa_sync_allpools(void)
+		(void) taskq_dispatch(spa->spa_sync_tq, spa_sync_tq_assign,
+		    &stq[i], TQ_FRONT);
+	}
+
+	/* wait on all syncthreads to start */
+	for (int i = 0; i < nthreads; i++) {
+		mutex_enter(&stq[i].sta_lock);
+		while (stq[i].sta_ready == 0)
+			cv_wait(&stq[i].sta_cv, &stq[i].sta_lock);
+		mutex_exit(&stq[i].sta_lock);
+	}
+
+	/* let all syncthreads resume, finish */
+	for (int i = 0; i < nthreads; i++) {
+		mutex_enter(&stq[i].sta_lock);
+		stq[i].sta_ready = 2;
+		cv_broadcast(&stq[i].sta_cv);
+		mutex_exit(&stq[i].sta_lock);
+	}
+	taskq_wait(spa->spa_sync_tq);
+
+	spa_taskqs_t *tqs =
+	    &spa->spa_zio_taskq[ZIO_TYPE_WRITE][ZIO_TASKQ_ISSUE];
+
+	spa_syncthread_info_t *ti = spa->spa_syncthreads;
+	for (int i = 0, w = 0; i < nthreads; i++, w++, ti++) {
+		ti->sti_thread = stq[i].sta_thread;
+		if (w == tqs->stqs_count) {
+			w = 0;
+		}
+		ti->sti_wr_iss_tq = tqs->stqs_taskq[w];
+		mutex_destroy(&stq[i].sta_lock);
+		cv_destroy(&stq[i].sta_cv);
+	}
+	kmem_free(stq, sizeof (spa_syncthread_info_t) * nthreads);
+
+	VERIFY(spa->spa_alloc_count == nthreads);
+	ti = spa->spa_syncthreads;
+	for (int a = 0; a < nthreads; a++, ti++) {
+		if (ti->sti_num_alloc == 0)
+			ti->sti_first_alloc = a;
+		ti->sti_num_alloc++;
+	}
+	return (spa->spa_sync_tq);
+}
+
+void
+spa_sync_tq_destroy(spa_t *spa)
+{
+	ASSERT(spa->spa_sync_tq != NULL);
+
+	taskq_wait(spa->spa_sync_tq);
+	taskq_destroy(spa->spa_sync_tq);
+	kmem_free(spa->spa_syncthreads,
+	    sizeof (kthread_t *) * spa->spa_alloc_count);
+	spa->spa_sync_tq = NULL;
+}
+
+void
+spa_select_allocator(zio_t *zio)
+{
+	zbookmark_phys_t *bm = &zio->io_bookmark;
+	spa_t *spa = zio->io_spa;
+
+	VERIFY(zio->io_type == ZIO_TYPE_WRITE);
+
+	/*
+	 * A gang block (for example) may have inherited its parent's
+	 * allocator, in which case there is nothing further to do.
+	 */
+	if (zio->io_flags & ZIO_FLAG_ALLOCATOR)
+		return;
+
+	ASSERT(spa != NULL);
+	ASSERT(bm != NULL);
+
+	zio->io_flags |= ZIO_FLAG_ALLOCATOR;
+
+	/*
+	 * We want to try to use as many allocators as possible to help improve
+	 * performance, but we also want logically adjacent IOs to be physically
+	 * adjacent to improve sequential read performance. We chunk each object
+	 * into 2^20 block regions, and then hash based on the objset, object,
+	 * level, and region to accomplish both of these goals.
+	 */
+	uint64_t hv = cityhash4(bm->zb_objset, bm->zb_object, bm->zb_level,
+	    bm->zb_blkid >> 20);
+
+	/*
+	 * First try to use an allocator assigned to the syncthread. Also
+	 * sets the write issue taskq for the allocator, if any.
+	 * Note, there must be an open pool to do this.
+	 */
+	if (spa->spa_sync_tq != NULL) {
+		spa_syncthread_info_t *ti = spa->spa_syncthreads;
+		for (int i = 0; i < spa->spa_alloc_count; i++, ti++) {
+			if (ti->sti_thread == curthread) {
+				zio->io_allocator =
+				    ((uint_t)hv % ti->sti_num_alloc) +
+				    ti->sti_first_alloc;
+				zio->io_wr_iss_tq = ti->sti_wr_iss_tq;
+				return;
+			}
+		}
```

Comment:
As I written above, the io_allocator is now just equal to thread, "i" here.  We do not even need the hash calculation here, which can be moved after this "if".

With all the 1:1 mapping we loose the "bm->zb_blkid >> 20" part of the hash logic, but hopefully it is not a huge deal.  Alternatively we could use the number of the thread as initial allocator to be then bumped for zb_layer >0 or zb_blkid >> 20, but I am not sure whether it worth it.  May think about it.

Reply:
fixed, thanks